### PR TITLE
Refactor.

### PR DESCRIPTION
**Pull request by bjlittle**
@JohannesBuchner Okay as promised ...

This PR is a bit of a bigger bang than I anticipated ... but essentially it includes the following:
* Enables travis-ci testing on all PRs. You will need to login to https://travis-ci.org/ and enable testing on your imagehash repo please
* Enables test code coverage. You will need to login to https://coveralls.io/ and enable testing on your imagehash repo please
* I've added a bit more capability to the `setup.py` so that now you can perform testing from the command line using `python setup.py test`
* As a result, I've move the tests into `imagehash/tests`
* I've split the tests (with the same capability) into separate test classes, so that we can grow the test coverage over each of the different algorithms
* To allow the tests to run without an internet connection, I've banked your lovely Alyson Hannigan test image within the repo
* I've added a pep8 test, that automatically ensures that all the code is [pep8](https://www.python.org/dev/peps/pep-0008/) compliant (apologies for the format changes, but that's a healthy change!)
* Can I convince you to adopt a [pep440](https://www.python.org/dev/peps/pep-0440/) styple of versioning your code?
* As you might be able to tell from travis-ci, using the PR, imagehash is running in python2.7, python3.4 and python3.5! Hoo-rah! :smile:

### CODE REVIEW

File: imagehash/__init__.py [hunk 1]
```diff
@@ -33,211 +33,243 @@
 
 from PIL import Image
 import numpy
-#import scipy.fftpack
-#import pywt
+
+
+__version__ = '3.1.0.dev0'
```
Comment (bjlittle) [inline imagehash/__init__.py]: This is pep440 versioning ...

File: imagehash/__init__.py [hunk 2]
```diff
@@ -33,211 +33,243 @@
...
 
 
 def _binary_array_to_hex(arr):
```
Comment (bjlittle) [inline imagehash/__init__.py]: All of the following changes (apologies) are for pep8 compliance (which is a very good thing indeed!)

File: imagehash/tests/__init__.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import os
+import os.path
+import unittest
+
+from PIL import Image
+
+import imagehash
+
+
+class TestImageHash(unittest.TestCase):
```
Comment (bjlittle) [inline imagehash/tests/__init__.py]: This is testing base class for common testing methods. Simply subclass to take advantage of this convenience.

File: imagehash/tests/test_average_hash.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+from __future__ import (absolute_import, division, print_function)
+
+from PIL import Image
+import unittest
+
+import imagehash
+import imagehash.tests as tests
+
+
+class Test(tests.TestImageHash):
```
Comment (bjlittle) [inline imagehash/tests/test_average_hash.py]: Note the sub-classing of our lovely new testing class.

File: setup.py [hunk 5]
```diff
@@ -1,28 +1,77 @@
...
-    name='ImageHash',
-    version='3.0',
+NAME = 'imagehash'
+DESCRIPTION = 'Image Hashing Library'
+DIR = os.path.abspath(os.path.dirname(__file__))
+
+
+package_data = {
+    NAME: [os.path.join('tests', 'data', '*')],
+}
```
Comment (bjlittle) [inline setup.py]: Package the tests with the code, so that users can also test the deployed code base

File: setup.py [hunk 6]
```diff
@@ -1,28 +1,77 @@
...
+
+def extract_packages():
+    packages = []
+    root = os.path.join(DIR, NAME)
+    offset = len(os.path.dirname(root)) + 1
+    for dpath, dnames, fnames in os.walk(root):
+        if os.path.exists(os.path.join(dpath, '__init__.py')):
+            package = dpath[offset:].replace(os.path.sep, '.')
+            packages.append(package)
+    return packages
```
Comment (bjlittle) [inline setup.py]: Discover all the code for `imagehash` that we care about and want to deploy

File: setup.py [hunk 7]
```diff
@@ -1,28 +1,77 @@
...
-        "pillow",      # or PIL
-        "PyWavelets",  # for whash
+        'six',         # for testing
+        'pep8',        # for testing
+        'numpy',
+        'scipy',       # for phash
+        'pillow',      # or PIL
+        'PyWavelets',  # for whash
     ],
+    test_suite = '{}.tests'.format(NAME),
```
Comment (bjlittle) [inline setup.py]: Allows us to `python setup.py test` :wink:

File: test_whash.py [hunk 8]
```diff
@@ -1,38 +0,0 @@
...
-		hash_size = 16
-		hash = imagehash.whash(image, hash_size=hash_size, image_scale=64)
-		self.assertEqual(hash.hash.size, hash_size**2)
-
-	def test_hash_size_more_than_scale(self):
-		image = self.get_white_image( (512, 512) )
-		self.assertRaises(AssertionError, imagehash.whash, image, hash_size=32, image_scale=16)
-
-if __name__ == '__main__':
-    unittest.main()
```
Comment (bjlittle) [inline test_whash.py]: All these tests have been refactored and relocated.

**Comment by JohannesBuchner** [comment]
Hi @bjlittle,

This pull-request contains several independent things and should be split up.

I think you are trying to do the testing foremost, so lets focus on that first, and move the other changes to another PR.

setup.py is getting too complicated, please keep it simple. We have one file, and one package, there is no reason to write parsing functions. Lets just keep the version in setup.py. I am ok with adding six as a dependency (is it possible to separate testing dependencies and install dependencies?).

> To allow the tests to run without an internet connection, I've banked your lovely Alyson Hannigan test image within the repo
That test image was used in one of the blog posts from which I started. I don't think you have the right to redistribute the image.

I think we discussed not changing the code style. Please remove such changes from this PR. We may later think about formalising a Codestyle check. To be honest I think this is overkill for this tiny project.

> Can I convince you to adopt a pep440 styple of versioning your code?
You can change the version in setup.py to -dev if you need to separate the pypi package from your current development version.

> As you might be able to tell from travis-ci, using the PR, imagehash is running in python2.7, python3.4 and python3.5! 
Great! Do you have a link of travis-ci testing against your fork?

**Comment by bjlittle** [comment]
@JohannesBuchner See https://travis-ci.org/bjlittle/imagehash/builds/170963008 and https://coveralls.io/github/bjlittle/imagehash

**Comment by bjlittle** [comment]
@JohannesBuchner Okay, I'll break this PR down and reduce the scope, thanks

**Pull request closed without merging**